File: competitive_programming/leetcode/06_javascript/202_happy_number.js
function happyNumber(n) {
  const seen = new Set();
  let value = n;
  while (value !== 1 && !seen.has(value)) {
    seen.add(value);
    value = String(value).split("").reduce((acc, digit) => acc + Number(digit) ** 2, 0);
  }
  return value === 1;
}

if (require.main === module) {
  if (happyNumber(19) !== true || happyNumber(2) !== false) throw new Error("happy number failed");
  console.log("202 happy number ok");
}

File: competitive_programming/leetcode/11_golang/202_happy_number.go
package main

import "fmt"

func happyNumber(n int) bool {
	seen := map[int]bool{}
	for n != 1 && !seen[n] {
		seen[n] = true
		next := 0
		for _, digit := range fmt.Sprintf("%d", n) {
			value := int(digit - '0')
			next += value * value
		}
		n = next
	}
	return n == 1
}

func main() {
	if !happyNumber(19) || happyNumber(2) {
		panic("happy number failed")
	}
	fmt.Println("202 happy number ok")
}
